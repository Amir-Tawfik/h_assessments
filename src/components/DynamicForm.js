import React, { useState } from 'react';

export default function DynamicForm(props) {

    ///// Constructor and initiations
    let roomsSet = [];
    var count;
    // retrieve saved data if submitted before and set room fields
    if (localStorage.getItem('roomsSavedData')) {
        const roomsSavedData = JSON.parse(localStorage.getItem('roomsSavedData'))
        roomsSet = roomsSet.concat(roomsSavedData);
        count = roomsSet.length;
    } else {
        // else 
        roomsSet = [{ roomNumber: 1, adultsCount: 1, childrenCount: 0, selected: true }]
        count = 1;
    }
    for (let num = count + 1; num <= props.roomsCount; num++) {
        roomsSet.push({ roomNumber: num, adultsCount: 1, childrenCount: 0, selected: false })
    }
    const [rooms, setRooms] = useState(roomsSet);


    ///// Handling functions
    const handleDropdownChange = roomNumber => event => {
        const name = event.target.name
        const value = event.target.value
        const index = rooms.findIndex(r => r.roomNumber === roomNumber)
        const newRooms = rooms.map((room, sidx) => {
            if (index !== sidx) return room;
            return { ...room, [name]: parseInt(value) };
        });
        setRooms(newRooms);
    }

    const handleCheckboxChange = roomNumber => event => {
        const checked = event.target.checked;
        const newRooms = rooms.map((room) => {
            if (checked === true && roomNumber < room.roomNumber) return room;
            if (checked === false && roomNumber > room.roomNumber) return room;
            return {
                ...room, selected: checked,
                adultsCount: checked ? room.selected ? room.adultsCount : rooms[0].adultsCount : props.adultOptions[0],
                childrenCount: checked ? room.selected ? room.childrenCount : rooms[0].childrenCount : props.childrenOptions[0]
            };
        });
        setRooms(newRooms);
    }

    const handleSubmit = (event) => {
        event.preventDefault();
        const roomsToSubmit = rooms.filter(r => r.selected)
        localStorage.setItem('roomsSavedData', JSON.stringify(roomsToSubmit));
    }

    const clearSavedData = (event) => {
        event.preventDefault();
        localStorage.removeItem('roomsSavedData');
        window.location.reload(false);
    }



    ///// HTML Contents
    const Checkbox = (props) => {
        return props.roomNumber > 1 ?
            <input id={"checkbox-" + props.roomNumber} checked={props.selected} onChange={handleCheckboxChange(props.roomNumber)} type="checkbox"></input> : ''
    }

    const SubmitField = () =>
        <fieldset className="submit-field">
            <input className="btn btn-primary" type="submit" value="Submit" />
            <button className="btn btn-default" onClick={clearSavedData}>Clear Saved Data</button>
        </fieldset>;

    return (
        <main>
            <form className="dynamic-form container" onSubmit={handleSubmit}>
                <section className="row">
                    {rooms.map((room, key) =>
                        <fieldset key={room.roomNumber} className="input-field">
                            <h6>
                                <Checkbox roomNumber={room.roomNumber} selected={room.selected}></Checkbox>
                                <label htmlFor={"checkbox-" + room.roomNumber} ><strong>Room {room.roomNumber}</strong></label>
                            </h6>
                            <div className={room.selected ? "row" : "disabled row"}>
                                <section className="col-sm-6">
                                    <label htmlFor={"adultsCount-" + room.roomNumber}>Adults (18+)</label>
                                    <select name="adultsCount" value={room.adultsCount} onChange={handleDropdownChange(room.roomNumber)}
                                        disabled={!room.selected} id={"adultsCount-" + room.roomNumber}>
                                        {props.adultOptions.map((number) => <option key={number} value={number}>{number}</option>)}
                                    </select>
                                </section>
                                <section className="col-sm-6">
                                    <label htmlFor={"childrenCount-" + room.roomNumber} >Children (0-17)</label>
                                    <select name="childrenCount" value={room.childrenCount} onChange={handleDropdownChange(room.roomNumber)}
                                        disabled={!room.selected} id={"childrenCount-" + room.roomNumber}>
                                        {props.childrenOptions.map((number) => <option key={number} value={number}>{number}</option>)}
                                    </select>
                                </section>
                            </div>
                        </fieldset>)}
                </section>
                <section className="row">
                    <SubmitField></SubmitField>
                </section>
            </form>
        </main>
    )
}
