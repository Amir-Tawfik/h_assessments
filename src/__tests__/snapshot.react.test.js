
import React from 'react';
import DynamicForm from '../components/DynamicForm';
import renderer from 'react-test-renderer';


test('Link changes the class when hovered', () => {
  const component = renderer.create(
    <DynamicForm roomsCount={6} childrenOptions={[0, 1, 2]} adultOptions={[1, 2]}> </DynamicForm>);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

});