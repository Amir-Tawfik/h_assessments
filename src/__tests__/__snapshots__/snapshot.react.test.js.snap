// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<main>
  <form
    className="dynamic-form container"
    onSubmit={[Function]}
  >
    <section
      className="row"
    >
      <fieldset
        className="input-field"
      >
        <h6>
          
          <label
            htmlFor="checkbox-1"
          >
            <strong>
              Room 
              1
            </strong>
          </label>
        </h6>
        <div
          className="row"
        >
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="adultsCount-1"
            >
              Adults (18+)
            </label>
            <select
              disabled={false}
              id="adultsCount-1"
              name="adultsCount"
              onChange={[Function]}
              value={1}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="childrenCount-1"
            >
              Children (0-17)
            </label>
            <select
              disabled={false}
              id="childrenCount-1"
              name="childrenCount"
              onChange={[Function]}
              value={0}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
        </div>
      </fieldset>
      <fieldset
        className="input-field"
      >
        <h6>
          <input
            checked={false}
            id="checkbox-2"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="checkbox-2"
          >
            <strong>
              Room 
              2
            </strong>
          </label>
        </h6>
        <div
          className="disabled row"
        >
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="adultsCount-2"
            >
              Adults (18+)
            </label>
            <select
              disabled={true}
              id="adultsCount-2"
              name="adultsCount"
              onChange={[Function]}
              value={1}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="childrenCount-2"
            >
              Children (0-17)
            </label>
            <select
              disabled={true}
              id="childrenCount-2"
              name="childrenCount"
              onChange={[Function]}
              value={0}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
        </div>
      </fieldset>
      <fieldset
        className="input-field"
      >
        <h6>
          <input
            checked={false}
            id="checkbox-3"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="checkbox-3"
          >
            <strong>
              Room 
              3
            </strong>
          </label>
        </h6>
        <div
          className="disabled row"
        >
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="adultsCount-3"
            >
              Adults (18+)
            </label>
            <select
              disabled={true}
              id="adultsCount-3"
              name="adultsCount"
              onChange={[Function]}
              value={1}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="childrenCount-3"
            >
              Children (0-17)
            </label>
            <select
              disabled={true}
              id="childrenCount-3"
              name="childrenCount"
              onChange={[Function]}
              value={0}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
        </div>
      </fieldset>
      <fieldset
        className="input-field"
      >
        <h6>
          <input
            checked={false}
            id="checkbox-4"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="checkbox-4"
          >
            <strong>
              Room 
              4
            </strong>
          </label>
        </h6>
        <div
          className="disabled row"
        >
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="adultsCount-4"
            >
              Adults (18+)
            </label>
            <select
              disabled={true}
              id="adultsCount-4"
              name="adultsCount"
              onChange={[Function]}
              value={1}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="childrenCount-4"
            >
              Children (0-17)
            </label>
            <select
              disabled={true}
              id="childrenCount-4"
              name="childrenCount"
              onChange={[Function]}
              value={0}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
        </div>
      </fieldset>
      <fieldset
        className="input-field"
      >
        <h6>
          <input
            checked={false}
            id="checkbox-5"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="checkbox-5"
          >
            <strong>
              Room 
              5
            </strong>
          </label>
        </h6>
        <div
          className="disabled row"
        >
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="adultsCount-5"
            >
              Adults (18+)
            </label>
            <select
              disabled={true}
              id="adultsCount-5"
              name="adultsCount"
              onChange={[Function]}
              value={1}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="childrenCount-5"
            >
              Children (0-17)
            </label>
            <select
              disabled={true}
              id="childrenCount-5"
              name="childrenCount"
              onChange={[Function]}
              value={0}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
        </div>
      </fieldset>
      <fieldset
        className="input-field"
      >
        <h6>
          <input
            checked={false}
            id="checkbox-6"
            onChange={[Function]}
            type="checkbox"
          />
          <label
            htmlFor="checkbox-6"
          >
            <strong>
              Room 
              6
            </strong>
          </label>
        </h6>
        <div
          className="disabled row"
        >
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="adultsCount-6"
            >
              Adults (18+)
            </label>
            <select
              disabled={true}
              id="adultsCount-6"
              name="adultsCount"
              onChange={[Function]}
              value={1}
            >
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
          <section
            className="col-sm-6"
          >
            <label
              htmlFor="childrenCount-6"
            >
              Children (0-17)
            </label>
            <select
              disabled={true}
              id="childrenCount-6"
              name="childrenCount"
              onChange={[Function]}
              value={0}
            >
              <option
                value={0}
              >
                0
              </option>
              <option
                value={1}
              >
                1
              </option>
              <option
                value={2}
              >
                2
              </option>
            </select>
          </section>
        </div>
      </fieldset>
    </section>
    <section
      className="row"
    >
      <fieldset
        className="submit-field"
      >
        <input
          className="btn btn-primary"
          type="submit"
          value="Submit"
        />
        <button
          className="btn btn-default"
          onClick={[Function]}
        >
          Clear Saved Data
        </button>
      </fieldset>
    </section>
  </form>
</main>
`;
